main: allow preloading conversation with -p and add -st / --single-turn

CISC opened the pull request:
Implements preloading conversation mode with `-p` and adds `-st`/`--single-turn` that limit the conversation to a single turn for one-shot instructions.

CISC committed:
Add chat template formatting to -no-cnv

CISC committed:
only enable prompt formatting if explicitly enabled

CISC commented:
@ngxson Ready for review.

ngxson commented:
IMO this feature can be quite confused and at the same time, introduce too much complication to the already complex `main.cpp` source code.

The problems are:
1. Traditionally, `-no-cnv` means there is no chat template at all. Now adding this, and we break that implication.
2. How can user know that this is a thing? We don't document it anywhere, no example, etc.
3. I'm not sure if anyone even needs this. Basically, most simple use case would be someone what to test the model by doing a single turn chat. The problem is, now user need to know the chat template to be used in advance. This is not a good UX.

CISC commented:
>     1. Traditionally, `-no-cnv` means there is no chat template at all. Now adding this, and we break that implication.

And it still does.

>     2. How can user know that this is a thing? We don't document it anywhere, no example, etc.

The user surely expects the chat template to be used if they enabled --jinja or --chat-template.

>     3. I'm not sure if anyone even needs this. Basically, most simple use case would be someone what to test the model by doing a single turn chat. The problem is, now user need to know the chat template to be used in advance. This is not a good UX.

I find this extremely useful as you can now do a one-shot prompt without having to paste a completely formatted template in your prompt (and if you really need to do that, it still works like that if you haven't explicitly enabled a chat template).

ngxson commented:
> > ```
> > 1. Traditionally, `-no-cnv` means there is no chat template at all. Now adding this, and we break that implication.
> > ```
> 
> 
> And it still does.

Versus what you said on the description:

> Adds chat template formatting to -no-cnv

Could you elaborate?

> The user surely expects the chat template to be used if they enabled --jinja or --chat-template.

In the case of one-shot prompt, user also expect to use the built-in template inside the model.

For example, a normal user will expect:

```
llama-cli -sys "You are Bob"
> who are you?
```

and

```
llama-cli -sys "You are Bob" -p "who are you?"
```

To be equivalent. There is no where in the document or in the `--help` message indicate that I must add `--jinja` or `--chat-template`, so I don't even know! That's what make it confused.

Simple solution is as follow:
- If `-p` is provided in conversation mode, we use the provided prompt to pre-fill the first user message in the conversation. This is kinda equivalent to `?q=...` parameter on chatgpt where it pre-fill the provided param into the input field.
- We add another flag called `-st, --single-turn`. If is set, we just exit after the first message.

So go back the my example above:

```
llama-cli -sys "You are Bob" -p "who are you?"
> who are you?
I am Bob
> ... (I can still ask more questions here)
```

and with `--single-turn`

```
llama-cli -sys "You are Bob" -p "who are you?" --single-turn
> who are you?
I am Bob
(exit with code 0)
```

CISC commented:
I see, so your objection is the `explicit_chat_template` check I added? I too would prefer it just to be enabled by default, but since that would break old behavior (and remove the ability to send a fully formatted prompt) I added the check.

Would the better solution be to warn users about the new behavior when automatically using chat template, and maybe add a raw prompt option?

CISC commented:
>     * We add another flag called `-st, --single-turn`. If is set, we just exit after the first message.

If I understand you correctly this would obsolete `-no/-cnv` params, `-st` would behave like `-no-cnv` if `-p` was set, and a single turn in conversation mode if not. Another PR? :)

ngxson commented:
No, nothing is obsolete.

-no-cnv == don't use chat template at all
--single-turn == conversation mode (-cnv) with  single turn

The -no-cnv mode never has support for chat template

CISC commented:
> The -no-cnv mode never has support for chat template

Yeah, I know, but what's the reason for keeping it like that? Conversation mode in itself never meant it was using a chat template, this is fairly recent. Even defaulting to conversation mode is new. To me it makes sense to (soft) deprecate -no/-cnv if you introduce `--single-turn`.

ngxson commented:
> Yeah, I know, but what's the reason for keeping it like that?

**For the technical reason:**

The -no-cnv and -cnv currently shares the same system under the hood. This is because historically, there was no -cnv mode, it was added with initial support for chatml template only. The support for other templates is added after that.

So to me, the template system is tightly coupled to the -cnv mode, and it should always be the case.

One big problem we're currently facing with chat templates is that after the introduction of jinja, sometimes it does not work correctly because turns are tracked at character level, not token level. I tried to address this by completely separate the -cnv into its own class: https://github.com/ggml-org/llama.cpp/pull/11203 - but that's the idea, the PR is not merge-able, but at some point in the future we will have completely different code base for the -cnv mode.

**For the UX:**

So from your suggestion, I assume that these 2 commands are the same:

```
llama-cli -sys "You are Bob" --jinja
> who are you?

llama-cli -sys "You are Bob" -p "who are you?" --jinja
```

But then, why these 2 command below are NOT the same?

```
llama-cli -sys "You are Bob"
> who are you?
```

And

```
llama-cli -sys "You are Bob" -p "who are you?"
```

Then there is **nowhere** in the document or in the help message say that I must add `--chat-template` or `--jinja` to make this work. And even if it's mentioned somewhere in the documentation, it's not obvious for end-user.

ngxson commented:
> Now, since you suggested the `--single-turn` param, my suggestion is that we can do away with -no/-cnv completely as `--single-turn -p "..."` then would equal `-no-cnv -p "..."`. Adding f.ex. `--raw-prompt` would solve the need for passing a preformatted prompt.

This seems to be a breaking change for me. AFAIU from what you're saying, if I already using this command:

```
llama-cli -m ... -p "The meaning of life is" -no-cnv
```

Then after this change, I have to use `--raw-prompt`

```
llama-cli -m ... -p "The meaning of life is" -no-cnv --raw-prompt
```

A such breaking change like this is not expected for many users.

CISC commented:
> This seems to be a breaking change for me.

Yes, it would be breaking if we go for this, but no more breaking than making conversation mode default. 

This PR as-is right now is non-breaking.

> AFAIU from what you're saying, if I already using this command:
> [...]
> A such breaking change like this is not expected for many users.

Certainly, but in order to fix all the inconsistencies something has to get broken. 

> Also, could you add some concrete examples? I have very hard time understand what are proposing.

Ok, first let me outline how the PR as-is works right now...

Works exactly like before:
```
llama-cli -m ... -p "The meaning of life is" -no-cnv
```

Uses built in chat template:
```
llama-cli -m ... -p "What is the meaning of life?" -no-cnv --jinja
```

Uses externally provided chat template:
```
llama-cli -m ... -p "What is the meaning of life?" -no-cnv --chat-template-file ...
```

Now, we can either go for this and continue to accept that -no-cnv behaves differently unless explicitly told to use chat template vs -cnv automatically selecting chat template, or we can accept status quo where --jinja and --chat-template params have no effect on -no-cnv, or we can do as follows...

Adding --single-turn and --raw-prompt basically blends the concept of completion/preformatted/conversation into one, allowing you to do them all with consistent params, like so...

One-shot sentence completion:
```
llama-cli -m ... --raw-prompt "The meaning of life is" --single-turn
```

One-shot instruction with built in chat template:
```
llama-cli -m ... -p "What is the meaning of life?" --single-turn
```

One-shot preformatted instruction with no chat template:
```
llama-cli -m ... --raw-prompt "<|im_start|>user\nWhat is the meaning of life?<|im_end|>\n<|im_start|>assistant\n" --single-turn
```

Pre-started conversation mode with built in chat template:
```
llama-cli -m ... -p "What is the meaning of life?"
```

All of the above need no warning and can be considered non-breaking. The following ones can be changed to work the new way by automatically moving prompt to raw_prompt and output a warning about the new params since we can check that -no-cnv was used...

One-shot sentence completion:
```
llama-cli -m ... -p "The meaning of life is" -no-cnv
```

One-shot preformatted instruction with no chat template:
```
llama-cli -m ... -p "<|im_start|>user\nWhat is the meaning of life?<|im_end|>\n<|im_start|>assistant\n" -no-cnv
```

ngxson commented:
> no more breaking than making conversation mode default.

At least, it's a breaking change that most people want to have - most users out there simple user `llama-cli` for conversation, so they expect it to behaves like `ollama run` command

---

Re. your first 3 commands:

```
llama-cli -m ... -p "What is the meaning of life?" -no-cnv [--jinja] [--chat-template]
```

It technically looks ok for me, but the problem is that without adequate documentations, no one will even know that this is possible. I think if we ever add this (aka merge the current PR), it will mostly to be used by you.

---

Re. `--raw-prompt`, it's now even more confused to me, I can't see how it different from `-no-cnv`. My suggestion for `--single-turn` is restricted to `-cnv` only. It cannot be used in `-no-cnv` mode. Something like `-no-cnv --single-turn` is invalid and will be ignored.

In my idea, your 4 commands will become:

One-shot sentence completion:

```
llama-cli -m ... -p "The meaning of life is" -no-cnv
```

One-shot instruction with built in chat template:

```
llama-cli -m ... -p "What is the meaning of life?" [-cnv] --single-turn
```

One-shot preformatted instruction with no chat template:

```
llama-cli -m ... -p "<|im_start|>user\nWhat is the meaning of life?<|im_end|>\n<|im_start|>assistant\n"
```

Pre-started conversation mode with built in chat template:

```
llama-cli -m ... -p "What is the meaning of life?" [-cnv]
```

No breaking change, and yet still support what you need. Exactly what you proposed in the end.

CISC commented:
> It technically looks ok for me, but the problem is that without adequate documentations, no one will even know that this is possible. I think if we ever add this (aka merge the current PR), it will mostly to be used by you.

I agree. :)

> My suggestion for `--single-turn` is restricted to `-cnv` only. It cannot be used in `-no-cnv` mode. Something like `-no-cnv --single-turn` is invalid and will be ignored.

My point exactly, once you introduce --single-turn it's no longer about conversation vs no-conversation, it's just a question of single-turn vs multi-turn and formatting.

> One-shot preformatted instruction with no chat template:
> 
> ```
> llama-cli -m ... -p "<|im_start|>user\nWhat is the meaning of life?<|im_end|>\n<|im_start|>assistant\n"
> ```

You forgot -no-cnv here I guess?

> No breaking change, and yet still support what you need. Exactly what you proposed in the end.

Yep, except since -cnv is essentially a no-op and -no-cnv just means single turn raw prompt, my suggestion is to soft deprecate them for clarity of what they actually do, but this is all semantics, I'll change this PR to what you suggest and we can discuss this further afterwards.

CISC committed:
add -st / --single-turn

CISC committed:
add --single-turn and -p in conversation mode

CISC commented:
@ngxson Ready for re-review (test failure looks like server outage again).

ochafik commented on examples/main/main.cpp:
```diff
@@ -276,22 +276,28 @@ int main(int argc, char ** argv) {
+
+            if (!params.prompt.empty()) {
+                // format and append the user prompt
+                prompt += chat_add_and_format("user", params.prompt);
```
Something like this:

```c++
auto simple_msg = [](const std::string & role, const std::string & content) {
  common_chat_msg msg;
  msg.role = role;
  msg.content = content;
  return msg;
};

common_chat_templates_inputs inputs;
if (!params.system_prompt.empty()) {
  inputs.messages.push_back(simple_msg("system", params.system_prompt));
}
if (!params.prompt.empty()) {
  inputs.messages.push_back(simple_msg("user", params.prompt));
}
auto prompt = common_chat_templates_apply(chat_templates.get(), inputs).prompt;
```

(not sure where / how to fit the waiting_for_first_input logic, leaving this up to you ;-))

CISC replied:
Thanks, it wasn't immediately obvious how to do this, as it turns out the solution was even simpler. :)

ochafik replied:
Looking at your updated code, it still calls chat_add_and_format at least twice (before finally calling common_chat_templates_apply ) which seems wasteful, could you maybe ditch it altogether and go along the lines of the code suggested above? (or explain why that wouldn't work?)

Also the line `prompt = params.system_prompt;` reads weird, prompt is a final prompt (w/ template applied / wrapped in all sorts of things), while system_prompt is a system message content, so assigning it to prompt feels wrong.

CISC replied:
Yes, it's (minimally) wasteful, but I did it for two (three) reasons; first of all it occurred to me that `common_chat_format_single` might perform actions in the future that I want to avoid having duplicated, and we need to add the conversation to chat_msgs as well (also, it's nice to get debug output of system prompt and user prompt separately, then combined).

The assignment was mainly to reduce the number of times I refer to params.system_prompt, but I can revert that for readability.

CISC committed:
fix -sys + -p

CISC committed:
reword warning

CISC committed:
small readability change and fix (long) outdated example usage

bandoti commented:
Re. -cnv and -no-cnv, I think it is important for us to step out of the implementation for a moment and consider that: _the user may enable or disable conversation mode._

At the moment the implementation dictates that the default is "-cnv" read from model, but the user should always be able to enable/disable that in the future.

There may be a strange scenario where users need to force conversation mode on a base model or something!

ngxson commented on examples/main/main.cpp:
```diff
@@ -372,6 +386,11 @@ int main(int argc, char ** argv) {
         params.interactive = true;
     }
 
+    if (params.single_turn && !params.prompt.empty()) {
+        params.interactive = false;
+        params.interactive_first = false;
+    }
```
Also maybe we can add a check somewhere to make sure `--single-turn` can only be used in `-cnv`?

```cpp
if (params.single_turn && !params.conversation_mode) {
    LOG_ERR("%s:--single-turn cannot be used without -cnv\n", __func__);
}
```

CISC replied:
Does it necessitate an error (or even a warning really)? I mean, in effect it's still a single turn.

CISC replied:
@ngxson I'm trying to think of a reason this should only apply to conversation mode and not to interactive as well? Do you have an example?

ngxson replied:
> Does it necessitate an error (or even a warning really)? I mean, in effect it's still a single turn.

The notion of "turn" only exists in conversation / interactive mode, so I think it make more sense to completely invalidate this flag in `-no-cnv` mode.

But it's a small details, OK not to add this check rn

> I'm trying to think of a reason this should only apply to conversation mode and not to interactive as well (ie, defaulting waiting_for_first_input to interactive_first (before it's set to true by conversation_mode)

Hmm OK then, tbh I don't even remember the exactly which branches of code is activated by `interactive`. Basically this feature changes too much since it was first added (a long time ago). In fact, I would appreciate if in near future, someone can refactor these 3 booleans into an enum:
- `interactive`
- `interactive_first`
- `conversation`

CISC replied:
Well, I changed it to apply only in conversation mode, we can revisit this later if need be.

CISC commented on examples/main/main.cpp:
```diff
@@ -276,22 +276,36 @@ int main(int argc, char ** argv) {
             if (!prompt.empty()) {
-                prompt = chat_add_and_format("system", prompt);
+                // format the system prompt (will use template default if empty)
+                chat_add_and_format("system", prompt);
```
Yes, agreed, fixed already. :)

ngxson commented:
> At the moment the implementation dictates that the default is "-cnv" read from model, but the user should always be able to enable/disable that in the future.

Yes I'm thinking about making `-cnv` to be the default (not depends on the model anymore), since it is pretty much expected nowadays that `llama-cli` run as chat by default.

`-no-cnv` can always be used to go back to the completion mode.

> There may be a strange scenario where users need to force conversation mode on a base model or something!

I'm thinking about throwing an error in this case, maybe smth like:

```
Error: This model does not have a chat template, it may not support chat. Try manually specifying a chat template via --chat-template
```

CISC commented:
> There may be a strange scenario where users need to force conversation mode on a base model or something!

Sure, but that would either just be interactive mode or an external chat template.

bandoti commented:
Yeah this all really evolved from the Alpaca/Llama2 days when there was no conversation mode at all and we had to specify the delimiters EOG, etc. directly!

And then conversation mode was added without the Jinja parser.

As far as I understand, the idea of keeping the "-cnv" flag was to allow the legacy template system to be forced and override the Jinja templates.

E.G. -cnv => --no-jinja

CISC committed:
only activate single turn in conversation mode
